### PR TITLE
feat(Marxan): Adds new service to generate Project Summaries after a run

### PR DESCRIPTION
**Pull request by KevSanchez**
Adds a new OutputProjectSummariesService that will generate a series of CSV files with summary info for all the scenarios runs in a given project, and store a zip with those files on the DB. This will be triggered after completing a Marxan run, on the API side at the end of the whole process.

### Overview

_Please write a description. If the PR is hard to understand, provide a quick explanation of the code._

### Designs

_Link to the related design prototypes (if applicable)_

### Testing instructions

_Please explain how to test the PR: ID of a dataset, steps to reach the feature, etc._

### Feature relevant tickets

_Link to the related task manager tickets_

---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

### CODE REVIEW

File: api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    const bestSolutionData = await this.getBestSolutionDataForScenarios(
+      scenarios,
+    );
+    const summedSolutionData = await this.getSummedSolutionDataForScenarios(
+      scenarios,
+    );
+
+    const workingDirectory = OutputProjectSummariesService.getProjectTempFolder(
+      projectId,
+    );
```
Comment (hotzevzl) [inline api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:60]: I think we discussed that this may be highly unlikely, but... wouldn't it be safer to use a unique directory each time this code is run?

Comment (hotzevzl) [reply api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:60]: if so, I would still do it in a separate "refinement" PR: for the moment I'd say this is ok, in the interest of moving the PR on, in any case.

Comment (hotzevzl) [reply api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:60]: also, it's not just a case of "if several of these project summaries are compiled more or less at the same time for a single project", but also (tin foil hat on), if anywhere else in the code someone adds _something_ later on that uses the same path structure for something completely unrelated, then boom.

tl;dr there are very good reasons for `mktemp` to be an essential tool ;)

File: api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    headers.push('puid');
+    for (const projectScenarioId of projectScenarioIds) {
+      const scenarioId = this.formatProjectScenarioId(projectScenarioId);
+      headers.push(`S${scenarioId}_best`);
+      headers.push(`S${scenarioId}_ssoln`);
+    }
+
+    const csvData = [];
+    csvData.push(headers);
+    for (const planningUnit of planningUnits.sort()) {
```
Comment (hotzevzl) [inline api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:116]: same as above - we don't want a lexicographical sorting but a numeric sorting here

File: api/apps/geoprocessing/test/steps/given-scenario-pu-data-exists.ts [hunk 3]
```diff
@@ -187,3 +188,40 @@ export const GivenScenarioPuData = async (
     };
   });
 };
+
+export const GivenProjectPuData = async (
```
Comment (hotzevzl) [inline api/apps/geoprocessing/test/steps/given-scenario-pu-data-exists.ts:192]: have you checked if we don't have these elsewhere? looking at what these fixtures do rings some kind of bells... but I'm not sure :/

**Review by hotzevzl — APPROVED** [review]
thanks @KevSanchez , I haven't _run_ this but overall it should do what it should.

just check my notes about the use of `.sort()` (unless lexicographical sorting is what you had in mind, but I don't think this should be the case).

File: api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+  private arrangeSummaryCSV(
+    planningUnits: number[],
+    scenarios: Scenario[],
+    bestSolutionData: ProjectScenarioBestSolutionMapping,
+    summedSolutionData: ProjectScenarioSummedSolutionMapping,
+  ): CsvFormatterStream<FormatterRow, FormatterRow> {
+    const projectScenarioIds = scenarios
+      .map((scenario) => scenario.projectScenarioId)
+      .sort();
```
[the remark this answers is not in the document]
Comment (KevSanchez) [reply api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:98]: I assumed that having the type number[] it would apply a sorting algorithm that makes sense for that, but yeah, you're right; this will fail for larger planningUnitIds

File: api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts (hunk 1 above)
Comment (KevSanchez) [reply api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts:60]: Hmm, I had the temp folders to be based on the projectId so that it could be more easily identified in case of debugging and looking at the files in staging/production, but you got a good point there.
I'll do it on a separate PR then.

File: api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/projects/output-project-summaries/output-project-summaries.service.ts]: yep, this being Yavascript, WAT is always around the corner: https://www.destroyallsoftware.com/talks/wat

to be fair, once the TS is transpiled to JS, types are erased so the `Array.prototype.sort()` function is necessarily (?) going to be dumber than we'd wish 🤷🏼 

another option would be to use _true_ JS typed arrays such as `UInt32Array`, for which the `.sort()` method would do a proper numeric sorting, but your fix is idiomatic and doesn't rely on creating new typed arrays out of primitive arrays.

**Comment by hotzevzl** [comment]
@KevSanchez sorry I had missed this yesterday:
https://github.com/Vizzuality/marxan-cloud/actions/runs/5586804904/jobs/10211303525?pr=1375#step:3:4125

we'll also need this:

```diff
diff --git a/api/api.Dockerfile b/api/api.Dockerfile
index 53ae8f93c..add17526e 100644
--- a/api/api.Dockerfile
+++ b/api/api.Dockerfile
@@ -5,7 +5,7 @@ ENV NAME marxan-api
 ENV USER $NAME
 ENV APP_HOME /opt/$NAME

-RUN apk add --no-cache bash
+RUN apk add --no-cache bash zip
 RUN addgroup $USER && adduser -s /bin/bash -D -G $USER $USER

 WORKDIR $APP_HOME
```

File: api/apps/geoprocessing/test/steps/given-scenario-pu-data-exists.ts (hunk 3 above)
Comment (KevSanchez) [reply api/apps/geoprocessing/test/steps/given-scenario-pu-data-exists.ts]: I didn't find anything equivalent at first glance, but I did notice there was some redundancy to the way I refactored the functions so that it would affect other tests suites that were already using the combined scenario + project Pu function. Now it should look better
